google/cloud/security/scanner/audit/bigquery_rules_engine.py

```diff
@@ -0,0 +1,284 @@
+            resource_ids = resource.get('resource_ids')
+
+            if not resource_ids or len(resource_ids) < 1:
+                raise audit_errors.InvalidRulesSchemaError(
+                    'Missing resource ids in rule {}'.format(rule_index))
+
+            dataset_id = rule_def.get('dataset_id')
+            special_group = rule_def.get('special_group')
+            user_email = rule_def.get('user_email')
+            domain = rule_def.get('domain')
+            group_email = rule_def.get('group_email')
+            role = rule_def.get('role')
+
+            is_any_none = (dataset_id is None) or (special_group is None) or\
+                          (user_email is None) or (domain is None) or\
+                          (group_email is None) or (role is None)
+
+            if is_any_none:
+                raise audit_errors.InvalidRulesSchemaError(
+                    'Faulty rule {}'.format(rule_def.get('name')))
+
+            rule_def_resource = bq_acls.BigqueryAccessControls(
+                escape_and_globify(dataset_id),
+                escape_and_globify(special_group),
+                escape_and_globify(user_email),
+                escape_and_globify(domain),
+                escape_and_globify(group_email),
+                escape_and_globify(role.upper()))
+
+            rule = Rule(rule_name=rule_def.get('name'),
+                        rule_index=rule_index,
+                        rules=rule_def_resource)
+
+            #resource_rules = self.resource_rules_map.get(rule_index)
+
+            if not self.resource_rules_map.get(rule_index):
+                self.resource_rules_map[rule_index] = rule
+
+    def get_resource_rules(self):
+        """Get all the resource rules for (resource, RuleAppliesTo.*).
+
+        Returns:
+            list: A list of ResourceRules.
+        """
+        resource_rules = []
+
+        for resource_rule in self.resource_rules_map:
+            resource_rules.append(self.resource_rules_map[resource_rule])
+
+        return resource_rules
+
+
+class Rule(object):
+    """Rule properties from the rule definition file.
+       Also finds violations.
+    """
+
+    def __init__(self, rule_name, rule_index, rules):
+        """Initialize.
+
+        Args:
+            rule_name (str): Name of the loaded rule
+            rule_index (int): The index of the rule from the rule definitions
+            rules (dict): The rules from the file
+        """
+        self.rule_name = rule_name
+        self.rule_index = rule_index
+        self.rules = rules
+
+    def find_policy_violations(self, bigquery_acl):
+        """Find BigQuery acl violations in the rule book.
+
+        Args:
+            bigquery_acl (BigqueryAccessControls): BigQuery ACL resource
+
+        Yields:
+            namedtuple: Returns RuleViolation named tuple
+        """
+        dataset_id_bool = True
+        special_group_bool = True
+        user_email_bool = True
+        domain_bool = True
```

Comment:
nit: can we choose a more clear naming for these variables?  Something like:
```is_domain_violated```, or the inverse ```is_domain_valid```

This will make a big difference in readability when it is used in the evaluation of ```should_raise_violation``` below.

```
is_dataset_id_violated
is_special_group_violated
is_user_email_bool_violated
is_domain_violated
is_group_email_violated
is_role_violated
```